### ISSUE TITLE
Integration testing for the Chrome extension

### ISSUE THREAD

**Issue by paulirish** [GoogleChrome/lighthouse#4479]
Having this would help prevent bugs like #4466 #4467 #4469

Current best idea to pull this off:

1. Use puppeteer to drive chrome headful
1. Probably use some aspects of [this guide from eric](https://stackoverflow.com/questions/48089670/detect-and-test-chrome-extension-using-puppeteer)
1. Issue [`Input.dispatchKeyEvent`](https://chromedevtools.github.io/devtools-protocol/tot/Input#method-dispatchKeyEvent) and set a `nativeVirtualKeyCode`. Joel believes this sends the keys to browser and thus we could tab to the pageaction button. (perhaps puppeteer handles the details, if not we can fall back to `_client.send()` ourselves)
1. Assuming we can trigger and interact with the popup, we can test the entire flow. If we cannot, we can at least interact with the background page and trigger the run from there.
1. Assert the report page opened, has the right details, etc.

**Comment by wardpeet** [comment]
I can pick up this one :)

**Comment by ebidel** [comment]
I'd be willing to pick this up too. We need a good example of crx testing using pptr!

If the dispatch key stuff doesn't work out, you could try opening the page directly and test them:

``js
// Can we navigate to a chrome-extension page? YES!
const page = await browser.newPage();
// Create a stable id: https://stackoverflow.com/a/23877974/274673
await page.goto('chrome-extension://ipfiboohojhbonenbbppflmpfkakjhed/popup.html');
// click buttons, check boxes, etc.
```

...Although the extension would need to be modified a bit to accept a test url from an `env` var or something. The [extension looks at the current tab](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-extension/app/src/popup.js#L228) for the test url, and if the tab is the popup page, you're not going to have a test URL :(

`_client.send()`...we've got https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#class-cdpsession now.

**Comment by wardpeet** [comment]
Feel free to take it if you have the time for it :)

**Comment by ebidel** [comment]
Yea :\ Probably best if someone else tackles it. I've got too much going on right now. I'm happy to consult or review PRs. The SO answer should mostly get you cooking.

**Comment by paulirish** [comment]
After discussing with the pptr team (and a small prototype) it doesn't appear we can open the pop-up.html w

**Comment by paulirish** [comment]
After discussing with the pptr team (and a small prototype) it doesn't appear we can open the popup.html with the browserAction icon via the protocol/pptr.

Dunno if opening the page directly means it can talk to the BG page. If it does, let's do that.

The fallback plan is to kick off a run via the background page. 

The rest _should_ work.

The pptr work will need to be clever with targets and maybe execution contexts.

**Comment by wardpeet** [comment]
Ok I got it to work, puppeteer calls the runLighthouse.. and does an extension run but what should we test in this run? Are we happy with just having the report pop open so we had a run or should we check for if audits did not threw an error?

**Comment by patrickhulce** [comment]
I'm cool if we just assert that the report opened and some basic elements exist 👍

**Comment by wardpeet** [comment]
how do you see this getting integrated?
I'm now developing it on https://github.com/wardpeet/lighthouse-keeper/tree/lighthouse-extension should this be moved to lighthouse repo when it's done? Or should we install it through npm?

**Comment by patrickhulce** [comment]
Moving this to Lighthouse repo SGTM in a new test directory of lighthouse-extension perhaps?

**Comment by ebidel** [comment]
Feel free to cc me on the PR. I have a few suggestions.

**Comment by paulirish** [comment]
fixed by #4640
